records/2024-04-11.2FA-magic-link.md

```diff
@@ -0,0 +1,135 @@
+
+## Context
+
+Notify requires users to manually copy two factor authentication codes from their email then paste it into a text box on the site in order to log into Notify. This process is cumbersome and not that accessible, or easy to use.
+
+The aim of this initiative is to improve the accessibility and ease of use of Notify's login process.
+
+## Options
+
+### Increase the visibility of 2FA codes in emails
+The simplest approach is to increase the visiblility of codes in the email to make the code more obvious and easier to interact with:
+- Place the code on it's own line in the email
+- Bold or otherwise highlight the text
+- Increase font size
+
+While increasing visibility is an improvement, it does not address the cumbersome amount of actions needed to complete the login:
+
+```mermaid
+flowchart LR
+    subgraph Email Inbox Tab
+        id1(Open email) --> id2(Find the code)--> id3(Highlight it)--> id4(Copy it)
+    end
+	subgraph Notify Tab
+        id0(Enter username and password)
+        id6(Select 2FA code text box)-->id7(Paste in the code)-->id8(Click Submit)
+    end
+    id0 --Tab Switch--> id1
+    id4 --Tab Switch--> id6
+    id8-->id9((X))
+
+```
+
+### Use magic links
+By using a magic link that the user can simply click on to complete the login process, we reduce the number of manual interactions required by the user to log in, thus making the process less obstructive and feel more natural.
+
+
+```mermaid
+
+flowchart LR
+	subgraph Notify Tab
+        id1(Enter username and password)
+        id4(User is logged into Notify)
+    end
+    subgraph Email Inbox Tab
+        id2(Find the magic link)-->id3(Click the link)
+    end
+    id1-->id2
+    id3-->id4
+    id4-->id5((X))
+
+```
+
+## Security Considerations
+While the addition of magic links provides users with a less cumbersome, more accessible and convenient login experience there are a number of security implications to leveraging them.
+
+### Email Impersonation
+A malicious actor could send a user, who just logged in, an email containing a magic link of their own. This link could redirect the user to a phishing site enabling the malicious actor to collect partial auth information from the user trying to log in, providing a window where the attacker could obtain unauthorized access to Notify.
+
+The attack surface, while not impossibly small, relies on a number of time limited circumstances to be aligned for this type of attack to succeed:
+
+1. The attacker needs to know the moment the user clicked the sign-in button
+2. The attacker needs to send their phishing email before the legitimate email is received to maximize their success -OR- they would need to execute a man-in-the-middle attack to modify the legitimate email's magic link to point to the malicious site.
+3. If the attacker succeeds in phishing and retrieving partial login information, they would also need access to the real magic link that we sent to the user in order to successfully log in.
```

Comment:
I think there was also a mention that a cookie needs to be present on the client side after username/password login and prior to click on the magic link. This prevents phishing attempt as it would require the attacker to also possess that cookie, or the username/pw along with mail inbox access of the target.

Reply:
Good point about the cookies. I've re-worked the security considerations section to, hopefully, reflect a more accurate assessment of the threats we discussed with the security team.